File: Persona/Form1.cs
using PERSONAS;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Persona
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void btnLimpiar_Click(object sender, EventArgs e)
        {
            txtNombre.Text = "";
            txtPater.Text = "";
            txtMater.Text = "";
            txtEmail.Text = "";
            txtTel.Text = "";
            Fecha.Value=DateTime.Now;
            labRFC.Visible = false;
            labRFC.Text="XXXX000000";
            adverEmail.Visible = false;
            adverFecha.Visible = false;
            adverNombre.Visible = false;
            adverPater.Visible = false;
            adverMat.Visible = false;
            adverTele.Visible = false;
        }

        private void btnCalcular_Click(object sender, EventArgs e)
        {
            Contacto c=new Contacto(txtTel.Text,txtEmail.Text,txtNombre.Text,txtPater.Text,txtMater.Text,Fecha.Value);
            labRFC.Text=c.RFC();
            labRFC.Visible=true;
            
        }

        private void txtMater_TextChanged(object sender, EventArgs e)
        {

            if (String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text))
            {
                adverMat.Visible = true;
            }
            else
            {
                adverMat.Visible = false;
                if (textos(txtMater.Text))
                {
                    adverMat.Visible = false;
                }
                else { adverMat.Visible = true; }
            }
            FechaChanged();
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
                
                if (String.IsNullOrEmpty(txtEmail.Text)| String.IsNullOrWhiteSpace(txtEmail.Text)|
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text)|
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text)|
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text)|
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text)
                    ) {
                    btnCalcular.Enabled = false; }
                else {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }
        }

        private void txtNombre_TextChanged(object sender, EventArgs e)
        {
            if (String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text)) {
                adverNombre.Visible = true;
            }
            else
            {
                adverNombre.Visible = false;
                string n = txtNombre.Text.Replace(" ", "");
                int cant = 0;
                if (textos(n))
                {
                    adverNombre.Visible = false;
                    for (int i = 0; i < txtNombre.Text.Length; i++)
                    {
                        if (txtNombre.Text.Substring(i, 1) == " ")
                        {
                            cant += 1;
                        }
                    }
                    if (cant > 2)
                    {
                        adverNombre.Visible = true;
                    }
                    else
                    {
                        adverNombre.Visible = false;
                    }
                }
                else
                {
                    adverNombre.Visible = true;
                }
            }
            FechaChanged();
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
                
                if (String.IsNullOrEmpty(txtEmail.Text) | String.IsNullOrWhiteSpace(txtEmail.Text) |
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text) |
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text) |
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text) |
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text)
                    )
                {
                    btnCalcular.Enabled = false;
                }
                else
                {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }
        }
        private bool textos(string a) {
            for (int i=0;i<a.Length;i++) {
                if (Convert.ToChar(a.Substring(i, 1)) <= 90 & Convert.ToChar(a.Substring(i, 1)) >= 65)
                {

                }
                else if (Convert.ToChar(a.Substring(i, 1)) <= 122 & Convert.ToChar(a.Substring(i, 1)) >= 97)
                {

                }
                else if (a.Substring(i, 1) =="á"| a.Substring(i, 1) == "é" | a.Substring(i, 1) == "í" | a.Substring(i, 1) == "ó" |
                    a.Substring(i, 1) == "ú" | a.Substring(i, 1) == "ü" | a.Substring(i, 1) == "Á" | a.Substring(i, 1) == "É" |
                    a.Substring(i, 1) == "Í" | a.Substring(i, 1) == "Ó" | a.Substring(i, 1) == "Ü" | a.Substring(i, 1) == "Ú" 
                    | a.Substring(i, 1) == "ñ" | a.Substring(i, 1) == "Ñ")
                {

                }
                else { return false; }
            }
            return true;
        }

        private void txtPater_TextChanged(object sender, EventArgs e)
        {
            if (String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text))
            {
                adverPater.Visible = true;
            }
            else
            {
                adverPater.Visible = false;
                if (textos(txtPater.Text))
                {

                    adverPater.Visible = false;
                }
                else { adverPater.Visible = true; }
            }
            FechaChanged();
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
             
                if (String.IsNullOrEmpty(txtEmail.Text) | String.IsNullOrWhiteSpace(txtEmail.Text) |
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text) |
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text) |
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text) |
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text) 
                    )
                {
                    btnCalcular.Enabled = false;
                }
                else
                {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }
        }
        private void FechaChanged()
        {
            DateTime difer = Fecha.Value;
            if (difer.AddYears(18).CompareTo(DateTime.Now) > 0)
            {
                adverFecha.Visible = true;
            }
            else { adverFecha.Visible = false; }
        }
            private void Fecha_ValueChanged(object sender, EventArgs e)
        {
            DateTime difer = Fecha.Value;
            if (difer.AddYears(18).CompareTo(DateTime.Now) >0)
            {
                adverFecha.Visible = true;
            }
            else { adverFecha.Visible = false; }
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
                
                if (String.IsNullOrEmpty(txtEmail.Text) | String.IsNullOrWhiteSpace(txtEmail.Text) |
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text) |
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text) |
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text) |
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text) 
                    )
                {
                    btnCalcular.Enabled = false;
                }
                else
                {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }

        }

        private void txtEmail_TextChanged(object sender, EventArgs e)
        {
            adverEmail.Visible = false;
            if (String.IsNullOrEmpty(txtEmail.Text)) {
                adverEmail.Visible = true;
            }

            else {
                if (IsValidEmail(txtEmail.Text)) { }
                else
                {
                    adverEmail.Visible = true;
                }
                 }
            FechaChanged();
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
                
                if (String.IsNullOrEmpty(txtEmail.Text) | String.IsNullOrWhiteSpace(txtEmail.Text) |
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text) |
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text) |
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text) |
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text) 
                    )
                {
                    btnCalcular.Enabled = false;
                }
                else
                {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }
        }
        public bool IsValidEmail(string email)
        {
            if (string.IsNullOrWhiteSpace(email))
                return false;

            try
            {
                // Normalize the domain
                email = Regex.Replace(email, @"(@)(.+)$", DomainMapper,
                                      RegexOptions.None, TimeSpan.FromMilliseconds(200));

                // Examines the domain part of the email and normalizes it.
                string DomainMapper(Match match)
                {
                    // Use IdnMapping class to convert Unicode domain names.
                    var idn = new IdnMapping();

                    // Pull out and process domain name (throws ArgumentException on invalid)
                    string domainName = idn.GetAscii(match.Groups[2].Value);

                    return match.Groups[1].Value + domainName;
                }
            }
            catch (RegexMatchTimeoutException e)
            {
                return false;
            }
            catch (ArgumentException e)
            {
                return false;
            }

            try
            {
                return Regex.IsMatch(email,
                    @"^[^@\s]+@[^@\s]+\.[^@\s]+$",
                    RegexOptions.IgnoreCase, TimeSpan.FromMilliseconds(250));
            }
            catch (RegexMatchTimeoutException)
            {
                return false;
            }
        }
        private void txtTel_TextChanged(object sender, EventArgs e)
        {
            txtTel.Text.Replace(" ","");
            adverTele.Visible = false;
            if (txtTel.Text.Length == 10) {
                if (txtTel.Text.StartsWith("7")| txtTel.Text.StartsWith("8")| txtTel.Text.StartsWith("9")) { 
                    for (int i = 0; i < txtTel.Text.Length; i++) {
                        if (Convert.ToChar(txtTel.Text.Substring(i, 1)) <= 57 & Convert.ToChar(txtTel.Text.Substring(i, 1)) >= 48)
                        {

                        }
                        else {
                            adverTele.Visible = true;
                        }
                    }
                }
                else {
                    adverTele.Visible = true;
                }
            }
            else { adverTele.Visible = true; }
            FechaChanged();
            if (adverEmail.Visible == false & adverFecha.Visible == false & adverNombre.Visible == false & adverMat.Visible == false & adverPater.Visible == false & adverTele.Visible == false)
            {
                
                if (String.IsNullOrEmpty(txtEmail.Text) | String.IsNullOrWhiteSpace(txtEmail.Text) |
                    String.IsNullOrEmpty(txtTel.Text) | String.IsNullOrWhiteSpace(txtTel.Text) |
                    String.IsNullOrEmpty(txtNombre.Text) | String.IsNullOrWhiteSpace(txtNombre.Text) |
                    String.IsNullOrEmpty(txtMater.Text) | String.IsNullOrWhiteSpace(txtMater.Text) |
                    String.IsNullOrEmpty(txtPater.Text) | String.IsNullOrWhiteSpace(txtPater.Text) 
                    )
                {
                    btnCalcular.Enabled = false;
                }
                else
                {
                    btnCalcular.Enabled = true;
                }
            }
            else { btnCalcular.Enabled = false; }
        }

        //un numero de telefono valido son 10 digitos y el primero es 7 8 o 9
    }

}

File: Contactos/Form1.vb
Imports System.Globalization
Imports System.Text.RegularExpressions
Imports PersonasVB

Public Class Form1
    Private Sub btnLimpiar_Click(sender As Object, e As EventArgs) Handles btnLimpiar.Click
        txtNombre.Text = ""
        txtPater.Text = ""
        txtMater.Text = ""
        txtEmail.Text = ""
        txtTel.Text = ""
        Fecha.Value = DateTime.Now
        labRFC.Visible = False
        labRFC.Text = "XXXX000000"
        adverEmail.Visible = False
        adverFecha.Visible = False
        adverNombre.Visible = False
        adverPater.Visible = False
        adverMat.Visible = False
        adverTele.Visible = False
    End Sub

    Private Sub btnCalcular_Click(sender As Object, e As EventArgs) Handles btnCalcular.Click
        Dim c As Contacto = New Contacto(txtTel.Text, txtEmail.Text, txtNombre.Text, txtPater.Text, txtMater.Text, Fecha.Value)
        labRFC.Text = c.RFC()
        labRFC.Visible = True
    End Sub

    Private Sub TextBox3_TextChanged(sender As Object, e As EventArgs) Handles txtMater.TextChanged

        If (String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text)) Then

            adverMat.Visible = True

        Else

            adverMat.Visible = False
            If (textos(txtMater.Text)) Then

                adverMat.Visible = False

            Else
                adverMat.Visible = True
            End If
        End If
        FechaChanged()
        If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


            If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
                    String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
                    String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
                    String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
                    String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
                    ) Then
                btnCalcular.Enabled = False
            Else
                btnCalcular.Enabled = True
            End If

        Else
            btnCalcular.Enabled = False
        End If
    End Sub

    Private Sub txtNombre_TextChanged(sender As Object, e As EventArgs) Handles txtNombre.TextChanged
        If (String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text)) Then
            adverNombre.Visible = True

        Else

            adverNombre.Visible = False
            Dim n As String = txtNombre.Text.Replace(" ", "")
            Dim cant As Integer = 0
            If (textos(n)) Then

                adverNombre.Visible = False
                For i As Integer = 0 To txtNombre.Text.Length - 1 Step 1

                    If (txtNombre.Text.Substring(i, 1) = " ") Then
                        cant += 1
                    End If
                Next
                If (cant > 2) Then

                    adverNombre.Visible = True

                Else
                    adverNombre.Visible = False
                End If

            Else
                adverNombre.Visible = True
            End If
        End If
        FechaChanged()
        If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


            If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
                    String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
                    String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
                    String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
                    String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
                    ) Then

                btnCalcular.Enabled = False

            Else

                btnCalcular.Enabled = True
            End If

        Else
            btnCalcular.Enabled = False
        End If
    End Sub

    Private Sub txtPater_TextChanged(sender As Object, e As EventArgs) Handles txtPater.TextChanged
        If (String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)) Then

            adverPater.Visible = True

        Else

            adverPater.Visible = False
            If (textos(txtPater.Text)) Then


                adverPater.Visible = False

            Else
                adverPater.Visible = True
            End If
        End If
            FechaChanged()
                If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


                    If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
                    String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
                    String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
                    String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
                    String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
                    ) Then

                        btnCalcular.Enabled = False

                    Else

                        btnCalcular.Enabled = True
                    End If

                Else
                    btnCalcular.Enabled = False
                End If
    End Sub
    Private Sub Fecha_ValueChanged(sender As Object, e As EventArgs) Handles Fecha.ValueChanged
        Dim difer As DateTime = Fecha.Value
        If (difer.AddYears(18).CompareTo(DateTime.Now) > 0) Then

            adverFecha.Visible = True

        Else
            adverFecha.Visible = False
        End If

        If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


                If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
                String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
                String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
                String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
                String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
                ) Then

                    btnCalcular.Enabled = False

                Else

                    btnCalcular.Enabled = True
                End If

            Else
                btnCalcular.Enabled = False
            End If
    End Sub

    Private Sub txtEmail_TextChanged(sender As Object, e As EventArgs) Handles txtEmail.TextChanged
        adverEmail.Visible = False
        If (String.IsNullOrEmpty(txtEmail.Text)) Then
            adverEmail.Visible = True
        Else
            If (IsValidEmail(txtEmail.Text)) Then

            Else
                adverEmail.Visible = True
            End If
        End If
        FechaChanged()
        If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


            If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
            String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
            String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
            String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
            String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
            ) Then

                btnCalcular.Enabled = False

            Else

                btnCalcular.Enabled = True
            End If

        Else
            btnCalcular.Enabled = False
        End If
    End Sub
    Private Sub txtTel_TextChanged(sender As Object, e As EventArgs) Handles txtTel.TextChanged

        txtTel.Text.Replace(" ", "")
        adverTele.Visible = False
        If (txtTel.Text.Length = 10) Then
            If (txtTel.Text.StartsWith("7") Or txtTel.Text.StartsWith("8") Or txtTel.Text.StartsWith("9")) Then
                For i As Integer = 0 To txtTel.Text.Length - 1 Step 1
                    If (Asc(txtTel.Text.Substring(i, 1)) <= 57 And Asc(txtTel.Text.Substring(i, 1)) >= 48) Then

                    Else
                        adverTele.Visible = True
                    End If
                Next
            Else
                adverTele.Visible = True
            End If
        Else
        adverTele.Visible = True
        End If
        FechaChanged()
                If (adverEmail.Visible = False And adverFecha.Visible = False And adverNombre.Visible = False And adverMat.Visible = False And adverPater.Visible = False And adverTele.Visible = False) Then


                    If (String.IsNullOrEmpty(txtEmail.Text) Or String.IsNullOrWhiteSpace(txtEmail.Text) Or
                    String.IsNullOrEmpty(txtTel.Text) Or String.IsNullOrWhiteSpace(txtTel.Text) Or
                    String.IsNullOrEmpty(txtNombre.Text) Or String.IsNullOrWhiteSpace(txtNombre.Text) Or
                    String.IsNullOrEmpty(txtMater.Text) Or String.IsNullOrWhiteSpace(txtMater.Text) Or
                    String.IsNullOrEmpty(txtPater.Text) Or String.IsNullOrWhiteSpace(txtPater.Text)
                    ) Then

                        btnCalcular.Enabled = False

                    Else

                        btnCalcular.Enabled = True
                    End If

                Else
                    btnCalcular.Enabled = False
                End If
    End Sub
    Private Function textos(a As String) As Boolean
        For i As Integer = 0 To a.Length - 1 Step 1
            If (Asc(a.Substring(i, 1)) <= 90 And Asc(a.Substring(i, 1)) >= 65) Then

            ElseIf (Asc(a.Substring(i, 1)) <= 122 And Asc(a.Substring(i, 1)) >= 97) Then

            ElseIf (a.Substring(i, 1) = "á" Or a.Substring(i, 1) = "é" Or a.Substring(i, 1) = "í" Or a.Substring(i, 1) = "ó" Or
                    a.Substring(i, 1) = "ú" Or a.Substring(i, 1) = "ü" Or a.Substring(i, 1) = "Á" Or a.Substring(i, 1) = "É" Or
                    a.Substring(i, 1) = "Í" Or a.Substring(i, 1) = "Ó" Or a.Substring(i, 1) = "Ü" Or a.Substring(i, 1) = "Ú" Or
                    a.Substring(i, 1) = "ñ" Or a.Substring(i, 1) = "Ñ") Then

            Else
                Return False
            End If
        Next
        Return True

    End Function

    Private Sub FechaChanged()

        Dim difer As DateTime = Fecha.Value
        If (difer.AddYears(18).CompareTo(DateTime.Now) > 0) Then

            adverFecha.Visible = True

        Else
            adverFecha.Visible = False
        End If
    End Sub
    Public Function IsValidEmail(email As String) As Boolean

        If (String.IsNullOrWhiteSpace(email)) Then
            Return False
        End If

        Try

            Dim DomainMaper As MatchEvaluator = New MatchEvaluator(AddressOf DomainMapper)
            ' Normalize the domain
            email = Regex.Replace(email, "(@)(.+)$", DomainMaper,
                                      RegexOptions.None, TimeSpan.FromMilliseconds(200))


        Catch e As RegexMatchTimeoutException

            Return False

        Catch e As ArgumentException

            Return False


        End Try

        Try

            Return Regex.IsMatch(email, "^[^@\s]+@[^@\s]+\.[^@\s]+$", RegexOptions.IgnoreCase, TimeSpan.FromMilliseconds(250))

        Catch e As RegexMatchTimeoutException

            Return False
    End Try

    End Function
    Private Function DomainMapper(match As Match) As String

        '// Use IdnMapping class to convert Unicode domain names.
        Dim idn = New IdnMapping()

        ' // Pull out And process domain name (throws ArgumentException on invalid)
        Dim domainName As String = idn.GetAscii(match.Groups(2).Value)

        Return match.Groups(1).Value + domainName
    End Function
End Class
